pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/CovarianceAggregationFunction.java

```diff
@@ -0,0 +1,237 @@
+    switch (blockValSet.getValueType().getStoredType()) {
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+        return blockValSet.getDoubleValuesSV();
+      default:
+        throw new IllegalStateException(
+            "Cannot compute covariance for non-numeric type: " + blockValSet.getValueType());
+    }
+  }
+
+  @Override
+  public void aggregateGroupBySV(int length, int[] groupKeyArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    double[] values1 = getValSet(blockValSetMap, _expression1);
+    double[] values2 = getValSet(blockValSetMap, _expression2);
+    for (int i = 0; i < length; i++) {
+      setGroupByResult(groupKeyArray[i], groupByResultHolder, values1[i], values2[i], values1[i] * values2[i], 1L);
+    }
+  }
+
+  @Override
+  public void aggregateGroupByMV(int length, int[][] groupKeysArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    double[] values1 = getValSet(blockValSetMap, _expression1);
+    double[] values2 = getValSet(blockValSetMap, _expression2);
+    for (int i = 0; i < length; i++) {
+      for (int groupKey : groupKeysArray[i]) {
+        setGroupByResult(groupKey, groupByResultHolder, values1[i], values2[i], values1[i] * values2[i], 1L);
+      }
+    }
+  }
+
+  @Override
+  public CovarianceTuple extractAggregationResult(AggregationResultHolder aggregationResultHolder) {
+    CovarianceTuple covarianceTuple = aggregationResultHolder.getResult();
+    if (covarianceTuple == null) {
+      return new CovarianceTuple(0.0, 0.0, 0.0, 0L);
+    } else {
+      return covarianceTuple;
+    }
+  }
+
+  @Override
+  public CovarianceTuple extractGroupByResult(GroupByResultHolder groupByResultHolder, int groupKey) {
+    return groupByResultHolder.getResult(groupKey);
+  }
+
+  @Override
+  public CovarianceTuple merge(CovarianceTuple intermediateResult1, CovarianceTuple intermediateResult2) {
+    intermediateResult1.apply(intermediateResult2);
+    return intermediateResult1;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getIntermediateResultColumnType() {
+    return DataSchema.ColumnDataType.OBJECT;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getFinalResultColumnType() {
+    return DataSchema.ColumnDataType.DOUBLE;
+  }
+
+  @Override
+  public Double extractFinalResult(CovarianceTuple covarianceTuple) {
+    long count = covarianceTuple.getCount();
+    if (count == 0L) {
+      return DEFAULT_FINAL_RESULT;
+    } else {
+      double sumX = covarianceTuple.getSumX();
+      double sumY = covarianceTuple.getSumY();
+      double sumXY = covarianceTuple.getSumXY();
+      double popCov = (sumXY / count) - (sumX / count) * (sumY / count);
+      double besselCorrection = count / (count - 1);
+      if (_isSample) {
+        return popCov * besselCorrection;
+      }
+      return popCov;
+    }
```

Comment:
```suggestion
      double popCov = (sumXY / count) - (sumX / count) * (sumY / count);
      double besselCorrection = count / (count - 1);
      if (_isSample) {
        return popCov * besselCorrection;
      }
      return popCov;
    }
```
```suggestion
      if (_isSample) {
        return (sumXY / count) - (sumX / count) * (sumY / (count - 1));
      } else {
        return (sumXY / count) - (sumX / count) * (sumY / count);
      }
    }
```
Let's avoid the multiplication then divide

Reply:
This part should actually be `(sumXY / divBase) - (sumX / count) * (sumY / divBase)` as the the `sample_covariance = popCov * besselCorrection = popCov * n/(n-1)`. We did the multiplication then divide to provide better readability and avoid this confusion. IMO it should be fine as we only pay a little price once in final reduction?

Reply:
yes. try to avoid divide if possible since this is on the critical path

Reply:
Can we also take care of divide by zero error ? May be add a test where denominator is 0.0 and result is NaN

https://stackoverflow.com/questions/14137989/java-division-by-zero-doesnt-throw-an-arithmeticexception-why

Reply:
also I am no expert in JIT compiler but `(sumX / count) * (sumY / count)` can be rewritten as `sumX * sumY / (count * count)` which also avoid another divide.

Reply:
Done. Ty for the suggestions!